### ISSUE TITLE
Using petals for pre-training

### ISSUE THREAD

**Issue by odellus** [bigscience-workshop/petals#272]
Since there really isn't any difference between pre-training and fine tuning, I was thinking maybe petals should do some kind of fold@home type of project to replicate the results of [LLaMA](https://github.com/facebookresearch/llama) and make the weights available on huggingface.

Thoughts? Discord/slack channel that would be a better place to have this discussion?

**Comment by mryab** [comment]
Hi! To be very specific, there *is* a difference between pretraining and finetuning: in Petals, the weights of the original model stay untouched, but in pretraining, you need to update billions of parameters for today's LLMs. This involves a lot more compute costs, and also you need to find some way to synchronize the weights between nodes at the same pipeline stage.

In fact, [SWARM Parallelism](https://arxiv.org/abs/2301.11913), one of the works that Petals is based on, studies precisely the problem of pretraining large models with volunteers. We have a [repo](https://github.com/yandex-research/swarm) for it, but it's not yet ready for as widespread use as Petals, since we decided to focus on the inference setting first. Besides, training a large model will likely involve thousands of A100-days, which IMO is currently not feasible with the numbers of people that are interested in running such experiments for a long time (Petals userbase is a proxy for people with such interests)

**Comment by odellus** [comment]
So petals is just for inference? I thought I saw it was also for fine
tuning. Only weights I freeze when fine tuning are in the embedding layer.
Figured if Petals could do SFT then adding embedding layer weight updates
wouldn't be that dramatic a change.

**Comment by mryab** [comment]
> So petals is just for inference? I thought I saw it was also for fine tuning.
You are right, Petals supports *local* finetuning: that is, the original weights of BLOOM stay the same, you are simply training adapters or continuous prompts on top of it. These addons to the model have less than 1% of overall BLOOM parameters, which makes it possible to train them on your machine with a relatively weak GPU without affecting others.

> Only weights I freeze when fine tuning are in the embedding layer.
If you are doing it with Petals and you have no adapters/deep p-tuning/LORA, then you are likely not updating any parameters during finetuning :) All Transformer weights on the servers over the network remain constant over time

**Comment by odellus** [comment]
Thank you for your insight @mryab. That definitely clears things up. I hadn't used Petals in the past and seeing "fine tuning" without understanding the specifics at the top of the README made me think Petals could do full gradient updates on the shared models. Excited to see what the future brings for

**Comment by borzunov** [comment]
Hi @odellus,

@mryab is right - Petals only supports parameter-efficient fine-tuning that stores a relatively small number of trainable parameters locally on the client's machine. This is enough to adapt the large model to most downstream tasks and leaves the opportunity to use the public swarm as the common infrastructure for doing computations with fixed pretrained blocks.

However, before Petals, we indeed worked on decentralized training of models from scratch - check out the [hivemind](https://github.com/learning-at-home/hivemind) library and a few models we've trained with it:

- https://huggingface.co/blog/collaborative-training (Bengali BERT)
- https://training-transformers-together.github.io (a basic text-to-image model similar to DALL-E 1)
- https://huggingface.co/CALM (Arabic BERT)

The [SWARM](https://arxiv.org/abs/2301.11913) paper mentioned by @mryab is the most advanced method among the ones hivemind supports (thanks to model parallelism). However, we have not trained any practical models with this recent method yet.

If you're a developer, you can set up a training run using these technologies yourself :) You'll need some time to run multiple experiments, choose architecture and hyperparams, solve all the networking issues, and find the actual collaborators willing to contribute their GPU time.

**Comment by odellus** [comment]
Hi @borzunov. Thanks for commenting! This is great stuff.

> ... and find people willing to contribute the GPU time for training your model.

This is the one I was hoping huggingface or the bigscience community would be able to provide as I lack the deep industry connections needed to kick off a project like that. I'm a nobody with zero clout.

I just saw too many tweets talking about how many millions it would cost to train an open source version of the LLaMa weights with the open source LLaMa code and it occurred to me that this would be a good use case for a distributed deep learning open collaboration. Petals came to mind as probably the most widely adopted distributed deep learning platform. Didn't realize it specializes in PEFT of local LoRA models and inference only.

I imagine if bigscience or huggingface were to write a blog post on how they're using hivemind or SWARM to train an open LLaMa model and they're looking for volunteers to speed up the process, they'd have a bunch of people interested in high quality open LLMs willing to contribute compute.

**Comment by borzunov** [comment]
I am not aware of any plans to train a new model collaboratively on the HF or BigScience side. We (the hivemind & Petals team) don't have capacity to work on new training runs right now, since we're busy with Petals and other projects.

However, if anyone else commits to conduct the preliminary experiments, set up the training run, and maintain it, there is a chance that they will be interested to share this in social media (given that the collaborative training technology is co-authored by people from HF and BigScience).
